contents/shells/bash.md

```diff
@@ -0,0 +1,169 @@
+
+Furthermore, if the code we write is buggy, it throws exception and error messages are directed to error output, then our result file will be empty! To handle it properly, we can also redirect error output to a file `errors`, and verify it is empty(such that no error occurs during execution) before we execute `verify`.
+
+```bash
+#!/bin/bash
+./calculate < DataBundle > result 2> errors
+if [ -s errors ]
+then
+     ./verify < result
+else
+     echo "error occured during execution."
+fi
+```
+Another useful command is `|`, which chained linux commands together, as such the output of previous commands is passed as input to the next command.
+
+The following command make use of shell command [ps](http://man7.org/linux/man-pages/man1/ps.1.html) and [grep](http://man7.org/linux/man-pages/man1/grep.1.html). Suppose you want to check the status of process `p` running on your system. Thus you typed in `ps aux`, and all the processes are listed, but there are too many processes that you can't find `p`. As such, you can pass the result of `(ps) aux` to `grep 'p'` to capture process `p`'s status. Thus the command will be look like `ps aux | grep 'p'`.
+
+Input/output streams can easily be redirected, allowing information to be sent or received from files or other applications(commands and user programs). This can mean test data can be easily supplied or output captured. A more detailed introduction to I/O stream redirection can be found [here](https://www.digitalocean.com/community/tutorials/an-introduction-to-linux-i-o-redirection).
+
+## Why use shell?
+
+A common question many people ask is: Why type commands in a shell when we can do the same things using GUI applications? Here are some reasons:
+
+- **Less lag in remote connection**: A remote command-line shell usually requires less bandwidth, which results in less lag. It makes it much easier to perform tasks in a cloud server, such as when a developer is connecting from home.
+
+- **Defined Standards**: No matter what linux distribution you are running, the command line shell is always available and usage is almost the same. However, GUI tools might be different in different distributions, make it very hard to learn.
+
+- **More powerful in the hands of an expert user**: You will have complete control of every option in shell. GUIs tends to simplify things, giving the user fewer options. While you might find command-line difficult to grasp, eventual usage will familiarize yourself such that the command-line is far easier to use than trying to find options in the drop-down menus and alike of a GUI. 
+
+- **Task automation**: bash script allows us to automate the frequently performed operations, it is easy to use and can be executed on any Unix-like operating systems without any modifications.
```

Comment:
>- **Faster**: Often, performing a task via the shell is faster than doing the same via a GUI application because the shell doesn't get slowed down by other overheads such as GUI rendering. For example, when performing tasks on a remote server using a shell, the lag will be less.
>- **Uniform**: Unlike GUI apps whose availability and usage can vary between Linux distributions, the shell is always available and the usage is almost the same across different Linux distributions. 
>- **More powerful**: GUIs tends to simplify things, giving the user fewer options. With a shell, you will have complete control of every option. While the shell has a steep learning curve, once you are familiar with it, you can do more things more efficiently with a shell compared to GUI applications. For example, you can automate things with a shell, something not easy to do with a GUI application.